sdk/cosmos/azure-cosmos/azure/cosmos/_timeout_failover_retry_policy.py

```diff
@@ -0,0 +1,102 @@
+# The MIT License (MIT)
+# Copyright (c) 2017 Microsoft Corporation
+
+# Permission is hereby granted, free of charge, to any person obtaining a copy
+# of this software and associated documentation files (the "Software"), to deal
+# in the Software without restriction, including without limitation the rights
+# to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+# copies of the Software, and to permit persons to whom the Software is
+# furnished to do so, subject to the following conditions:
+
+# The above copyright notice and this permission notice shall be included in all
+# copies or substantial portions of the Software.
+
+# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+# SOFTWARE.
+
+"""Internal class for timeout failover retry policy implementation in the Azure
+Cosmos database service.
+"""
+from . import http_constants
+from ..cosmos.documents import _OperationType
+
+
+class _TimeoutFailoverRetryPolicy(object):
+
+    Max_retry_attempt_count = 1
+    Retry_after_in_milliseconds = 0
+
+    def __init__(self, connection_policy, global_endpoint_manager, *args):
+        self._max_retry_attempt_count = 120
+        self.current_retry_attempt_count = 0
+        self.retry_after_in_milliseconds = 1000
+        self.args = args
+
+        self.global_endpoint_manager = global_endpoint_manager
+        self.failover_retry_count = 0
+        self.connection_policy = connection_policy
+        self.request = args[0] if args else None
+        # clear previous location-based routing directive
+        if self.request:
+            self.request.clear_route_to_location()
+
+            # Resolve the endpoint for the request and pin the resolution to the resolved endpoint
+            # This enables marking the endpoint unavailability on endpoint failover/unreachability
+            self.location_endpoint = self.global_endpoint_manager.resolve_service_endpoint(self.request)
+            self.request.route_to_location(self.location_endpoint)
```

Comment:
not sure if we should be doing this in the init since we create this policy before running the request and would effectively be setting these flags for all requests passing through the retry utility.

Reply:
It looks like we do the same thing in other retry policies. Since the same code is being repeated in `ShouldRetry` I will remove that from the `init` method.

Reply:
It looks like this part is necessary.  If we do not include this, then we cannot mark the endpoints as unavailable for read later once a request fails.